src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -222,45 +234,64 @@ public UnfilteredRowIterator computeNext()
         /**
          * Returns the next available key contained by one of the keyRanges and selected by the queryController.
          * If the next key falls out of the current key range, it skips to the next key range, and so on.
-         * If no more keys acceptd by the controller are available, returns null.
+         * If no more keys acceptd by the controller are available, returns an empty list.
          */
-        private @Nullable PrimaryKey nextSelectedKeyInRange()
+        private List<PrimaryKey> nextSelectedKeyInRange()
         {
+            List<PrimaryKey> threadLocalNextKeys = nextKeys.get();
+            threadLocalNextKeys.clear();
             PrimaryKey key;
+
             do
             {
                 key = nextKeyInRange();
+
+                if (key == null)
+                    return Collections.emptyList();
             }
-            while (key != null && queryController.doesNotSelect(key));
-            return key;
+            while (queryController.doesNotSelect(key) || key.equals(lastKey));
+
+            lastKey = key;
+            threadLocalNextKeys.add(key);
+            return threadLocalNextKeys;
         }
 
         /**
+         * TODO: update this description...
```

Comment:
leftover

Reply:
Yup, need to update all of this when things settle...